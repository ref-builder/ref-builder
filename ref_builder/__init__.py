"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
