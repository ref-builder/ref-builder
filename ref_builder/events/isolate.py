from pydantic import UUID4

from ref_builder.events.base import EventData, Event, IsolateQuery
from ref_builder.utils import IsolateName


class CreateIsolateData(EventData):
    """The data for a :class:`CreateIsolate` event."""

    id: UUID4
    legacy_id: str | None
    name: IsolateName | None


class CreateIsolate(Event):
    """An event that creates an isolate for a specific OTU."""

    data: CreateIsolateData
    query: IsolateQuery


class DeleteIsolateData(EventData):
    """The data for a :class:`DeleteIsolate` event."""

    rationale: str


class DeleteIsolate(Event):
    """An isolate deletion event."""

    data: DeleteIsolateData
    query: IsolateQuery
