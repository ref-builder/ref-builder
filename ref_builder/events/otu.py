from pydantic import UUID4

from ref_builder.events.base import EventData, Event, OTUQuery
from ref_builder.models import Molecule
from ref_builder.plan import MonopartitePlan, MultipartitePlan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: MonopartitePlan | MultipartitePlan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery


class ExcludeAccessionData(EventData):
    """The data for a :class:`ExcludeAccession` event."""

    accession: str


class ExcludeAccession(Event):
    """An accession exclusion event.

    This event is emitted when a Genbank accession is not going to be allowed in the
    reference.
    """

    data: ExcludeAccessionData
    query: OTUQuery


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: MonopartitePlan | MultipartitePlan


class CreatePlan(Event):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery


class SetReprIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetReprIsolate(Event):
    """An event that sets the representative isolate for an OTU."""

    data: SetReprIsolateData
    query: OTUQuery
