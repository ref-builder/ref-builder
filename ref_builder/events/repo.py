from pydantic import UUID4

from ref_builder.events.base import EventData, Event, RepoQuery
from ref_builder.resources import RepoSettings


class CreateRepoData(EventData):
    """The data for a :class:`CreateRepo` event."""

    id: UUID4
    data_type: str
    name: str
    organism: str
    settings: RepoSettings


class CreateRepo(Event):
    """An event that creates a new repository.

    This event is always the first event in a repository's event log.
    """

    data: CreateRepoData
    query: RepoQuery
