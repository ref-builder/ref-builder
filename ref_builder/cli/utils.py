import sys
import warnings
from uuid import UUID

import click

from ref_builder.errors import InvalidInputError, PartialIDConflictError
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.repo import Repo
from ref_builder.utils import OTUDeletedWarning

pass_repo = click.make_pass_decorator(Repo)

UUID_STRING_LENGTH = 36


def get_otu_from_identifier(repo: Repo, identifier: str) -> OTUBuilder:
    """Return an OTU id from the repo if identifier matches a single OTU.

    The identifier can either be a stringified UUIDv4, a truncated portion
    of a UUID, a NCBI Taxonomy ID or an acronym associated with the OTU.

    Raise a PartialIDConflictError if >1 OTU is found.

    :param repo: the repository to be searched.
    :param identifier: a non-UUID identifier.
        Can be an integer Taxonomy ID, acronym or truncated partial UUID.
    :return: the UUID of the OTU or ``None``
    """
    otu_id = None

    if len(identifier) == UUID_STRING_LENGTH:
        try:
            otu_id = UUID(identifier)
        except ValueError:
            otu_id = None

    elif identifier.isnumeric():
        try:
            taxid = int(identifier)
        except ValueError:
            pass
        else:
            otu_id = repo.get_otu_id_by_taxid(taxid)

    else:
        if (otu_id := repo.get_otu_id_by_acronym(identifier)) is not None:
            return repo.get_otu(otu_id)

        try:
            otu_id = repo.get_otu_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError as e:
            click.echo(
                e,
                err=True,
            )

    if otu_id is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    with warnings.catch_warnings(
        category=OTUDeletedWarning, record=True
    ) as warning_list:
        otu_ = repo.get_otu(otu_id)

    if otu_ is None:
        if warning_list:
            click.echo("OTU has been deleted.", err=True)

        else:
            click.echo("OTU not found.", err=True)

        sys.exit(1)

    return otu_


def get_otu_isolate_ids_from_identifier(repo: Repo, identifier: str) -> (UUID, UUID):
    """Return an isolate ID from the repo if identifier matches a single isolate.

    Handles cases where the isolate cannot be found.
    """
    isolate_id = None

    if len(identifier) == 32:
        try:
            isolate_id = UUID(identifier)
        except ValueError:
            isolate_id = None

    else:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError:
            click.echo(
                "Partial ID segment must be at least 8 characters long.", err=True
            )
            sys.exit(1)

    if isolate_id is not None:
        if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is not None:
            if isolate_id in repo.get_otu(otu_id).isolate_ids:
                return otu_id, isolate_id

    click.echo("Isolate ID could not be found.", err=True)
    sys.exit(1)
