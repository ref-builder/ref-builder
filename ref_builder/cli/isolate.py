import sys
from pathlib import Path
from uuid import UUID

import click

from ref_builder.cli.utils import pass_repo
from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.isolate import (
    add_and_name_isolate,
    add_genbank_isolate,
    add_unnamed_isolate,
)
from ref_builder.otu.modify import delete_isolate_from_otu
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.repo import Repo, locked_repo
from ref_builder.utils import IsolateName, IsolateNameType


@click.group(name="isolate")
@click.pass_context
@path_option
def isolate(ctx: click.Context, path: Path) -> None:
    """Manage isolates."""
    ctx.obj = ctx.with_resource(locked_repo(path))


@isolate.command(name="create")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--taxid", type=int, required=True)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@pass_repo
def isolate_create(
    repo: Repo,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    taxid: int,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

        sys.exit(0)

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
            sys.exit(0)

        except RefSeqConflictError as e:
            click.echo(
                f"{e.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as e:
        click.echo(
            f"{e.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)


@isolate.command(name="delete")  # type: ignore
@click.argument("IDENTIFIER", type=str)
@pass_repo
def isolate_delete(repo: Repo, identifier: str) -> None:
    """Delete an isolate with a UUID corresponding to IDENTIFIER.

    IDENTIFIER is an unique isolate ID (>8 characters)
    """
    isolate_id = None
    try:
        isolate_id = UUID(identifier)
    except ValueError:
        pass

    if isolate_id is None:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except InvalidInputError:
            click.echo("Partial ID segment must be at least 8 characters long.", err=True)
            sys.exit(1)

        except PartialIDConflictError as e:
            click.echo(e, err=True)

    if isolate_id is None:
        click.echo("Isolate could not be found.", err=True)
        sys.exit(1)

    if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is None:
        click.echo(f"The containing OTU could not be found.", err=True)
        sys.exit(1)

    if (otu_ := repo.get_otu(otu_id)) is None:
        click.echo(f"OTU not found.", err=True)
        sys.exit(1)

    delete_isolate_from_otu(repo, otu_, isolate_id)

    click.echo("Isolate deleted.")
