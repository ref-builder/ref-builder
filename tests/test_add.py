import pytest
from pytest_mock import MockerFixture
from structlog.testing import capture_logs

from ref_builder.ncbi.client import NCBIClient
from ref_builder.ncbi.models import (
    NCBIGenbank,
    NCBIRank,
    NCBITaxonomy,
    NCBITaxonomyOtherNames,
)
from ref_builder.otu.isolate import (
    add_and_name_isolate,
    add_genbank_isolate,
    add_unnamed_isolate,
)
from ref_builder.repo import Repo
from ref_builder.services.otu import OTUService
from ref_builder.utils import IsolateName, IsolateNameType
from tests.fixtures.factories import (
    NCBIGenbankFactory,
    NCBISourceFactory,
    NCBITaxonomyFactory,
)


def create_mocked_otu_service(
    repo: Repo,
    mocker: MockerFixture,
    taxonomy: NCBITaxonomy,
    records: list[NCBIGenbank],
) -> OTUService:
    """Create an OTUService with a mocked NCBIClient.

    :param repo: repository to use
    :param mocker: pytest-mock fixture
    :param taxonomy: taxonomy record to return from fetch_taxonomy_record
    :param records: genbank records to return from fetch_genbank_records
    :return: OTUService with mocked NCBIClient
    """
    ncbi_client = mocker.create_autospec(NCBIClient, instance=True)
    ncbi_client.fetch_genbank_records = mocker.Mock(return_value=records)
    ncbi_client.fetch_taxonomy_record = mocker.Mock(return_value=taxonomy)
    return OTUService(repo, ncbi_client)


class TestCreateOTU:
    @pytest.mark.parametrize(
        ("taxid", "segment_count"),
        [
            (3426695, 2),
            (3240630, 1),
        ],
    )
    def test_ok(
        self,
        empty_repo: Repo,
        mocker: MockerFixture,
        ncbi_taxonomy_factory: type[NCBITaxonomyFactory],
        ncbi_genbank_factory: type[NCBIGenbankFactory],
        taxid: int,
        segment_count: int,
    ):
        """Test the the default, happy case works."""
        taxonomy = ncbi_taxonomy_factory.build(id=taxid, rank=NCBIRank.SPECIES)

        if segment_count == 1:
            records = [
                ncbi_genbank_factory.build(
                    source__taxid=taxid,
                    source__isolate="Isolate A",
                    accession="AB123456",
                )
            ]
        else:
            records = ncbi_genbank_factory.build_isolate(
                segment_count=segment_count,
                refseq=False,
                base_source=NCBISourceFactory.build(taxid=taxid, isolate="Isolate A"),
            )

        otu_service = create_mocked_otu_service(empty_repo, mocker, taxonomy, records)
        accessions = [r.accession for r in records]

        with empty_repo.lock():
            otu = otu_service.create(accessions)

        assert otu
        assert otu.taxid == taxid
        assert empty_repo.get_otu_by_taxid(taxid) == otu

    def test_no_accessions(self, empty_repo: Repo):
        """Test that creating an OTU with not accession fails with the expected log."""
        otu_service = OTUService(empty_repo, NCBIClient(False))

        with empty_repo.lock(), capture_logs() as logs:
            otu = otu_service.create([])

            assert otu is None
            assert any(
                "OTU could not be created to spec" in (log.get("event") or "")
                for log in logs
            )

    def test_duplicate_taxid(
        self,
        empty_repo: Repo,
        mocker: MockerFixture,
        ncbi_taxonomy_factory: type[NCBITaxonomyFactory],
        ncbi_genbank_factory: type[NCBIGenbankFactory],
    ):
        """Test that an OTU with the same taxid cannot be created."""
        taxid = 345184
        taxonomy = ncbi_taxonomy_factory.build(id=taxid, rank=NCBIRank.SPECIES)
        records = ncbi_genbank_factory.build_isolate(
            segment_count=2,
            refseq=False,
            base_source=NCBISourceFactory.build(taxid=taxid, isolate="Isolate A"),
        )

        otu_service = create_mocked_otu_service(empty_repo, mocker, taxonomy, records)
        accessions = [r.accession for r in records]

        with empty_repo.lock():
            otu = otu_service.create(accessions)

        assert otu
        assert otu.taxid == taxid

        with empty_repo.lock():
            duplicate_otu = otu_service.create(accessions)
            assert duplicate_otu is None

    def test_refseq_autoexclude(
        self,
        empty_repo: Repo,
        mocker: MockerFixture,
        ncbi_taxonomy_factory: type[NCBITaxonomyFactory],
        ncbi_genbank_factory: type[NCBIGenbankFactory],
    ):
        """Test that the superceded accessions included in RefSeq metadata are
        automatically added to the OTU's excluded accessions list.
        """
        taxid = 3158377
        taxonomy = ncbi_taxonomy_factory.build(id=taxid, rank=NCBIRank.SPECIES)

        # Create 6 RefSeq records with comments pointing to old accessions
        old_accessions = [f"EF546{800 + i}" for i in range(8, 14)]

        records = ncbi_genbank_factory.build_isolate(
            segment_count=6,
            refseq=True,
            base_source=NCBISourceFactory.build(taxid=taxid, isolate="Isolate A"),
        )

        # Update the records with proper RefSeq accessions and comments
        for i, (record, old_acc) in enumerate(
            zip(records, old_accessions, strict=False), start=4
        ):
            record.accession = f"NC_01031{i}"
            record.comment = (
                f"PROVISIONAL REFSEQ: This record has not yet been subject to final NCBI review. "
                f"The reference sequence is identical to {old_acc}."
            )

        otu_service = create_mocked_otu_service(empty_repo, mocker, taxonomy, records)
        accessions = [r.accession for r in records]

        with empty_repo.lock():
            otu = otu_service.create(accessions)

        assert otu
        assert otu.excluded_accessions == set(old_accessions)

    def test_acronym(
        self,
        empty_repo: Repo,
        mocker: MockerFixture,
        ncbi_taxonomy_factory: type[NCBITaxonomyFactory],
        ncbi_genbank_factory: type[NCBIGenbankFactory],
    ) -> None:
        """Test that the acronym pulled from NCBI is correct."""
        other_names = NCBITaxonomyOtherNames()
        other_names.acronym = ["KLV"]

        taxonomy = NCBITaxonomy(
            id=132477,
            name="Kasba virus",
            rank=NCBIRank.SPECIES,
            other_names=other_names,
            lineage=[],
        )

        records = [
            ncbi_genbank_factory.build(source__taxid=taxonomy.id, accession="AB123456")
        ]

        otu_service = create_mocked_otu_service(empty_repo, mocker, taxonomy, records)

        with empty_repo.lock():
            otu = otu_service.create(["AB123456"])

        assert otu is not None
        assert otu.acronym == "KLV"


class TestAddIsolate:
    def test_multipartite(self, precached_repo: Repo):
        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu_before = otu_service.create(["MF062136", "MF062137", "MF062138"])

        assert otu_before
        assert otu_before.accessions == {"MF062136", "MF062137", "MF062138"}
        assert len(otu_before.isolate_ids) == 1

        with precached_repo.lock():
            isolate = add_genbank_isolate(
                precached_repo,
                otu_before,
                ["MF062125", "MF062126", "MF062127"],
                ncbi_client,
            )

        assert isolate

        otu_after = precached_repo.get_otu(otu_before.id)

        assert otu_after
        assert otu_after.accessions == {
            "MF062125",
            "MF062126",
            "MF062127",
            "MF062136",
            "MF062137",
            "MF062138",
        }

        isolate_after = otu_after.get_isolate(isolate.id)

        assert isolate_after
        assert isolate_after.accessions == {
            "MF062125",
            "MF062126",
            "MF062127",
        }

    def test_genbank(self, precached_repo: Repo):
        """Test that add_genbank_isolate() adds an isolate with a correctly parsed
        name.
        """
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu_before = otu_service.create(isolate_1_accessions)

            assert otu_before
            assert otu_before.accessions == set(isolate_1_accessions)

            isolate = add_genbank_isolate(
                precached_repo, otu_before, isolate_2_accessions, ncbi_client
            )

            assert isolate

        otu_after = precached_repo.get_otu(otu_before.id)

        assert otu_after
        assert otu_after.accessions == set(isolate_1_accessions).union(
            set(isolate_2_accessions),
        )

        assert isolate == otu_after.get_isolate(isolate.id)
        assert isolate.accessions == set(isolate_2_accessions)
        assert isolate.name == IsolateName(
            IsolateNameType.ISOLATE,
            "Douglas Castle",
        )

    def test_ignore_name(self, precached_repo: Repo):
        """Test that add_unnamed_isolate() adds the isolate with a ``None`` name."""
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu_before = otu_service.create(isolate_1_accessions)

        assert otu_before
        assert otu_before.accessions == set(isolate_1_accessions)

        isolate_1 = otu_before.isolates[0]

        assert isolate_1
        assert isolate_1.accessions == set(isolate_1_accessions)

        with precached_repo.lock():
            isolate_2 = add_unnamed_isolate(
                precached_repo, otu_before, isolate_2_accessions, ncbi_client
            )

        assert isolate_2

        otu_after = precached_repo.get_otu(otu_before.id)

        assert otu_after
        assert otu_after.isolate_ids == {isolate_1.id, isolate_2.id}

        assert isolate_2 == otu_after.get_isolate(isolate_2.id)
        assert isolate_2.name is None
        assert isolate_2.accessions == set(isolate_2_accessions)

    def test_add_and_name_isolate(self, precached_repo: Repo):
        """Test that add_and_name_isolate() creates an isolate with the correct name."""
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu = otu_service.create(isolate_1_accessions)

        assert otu
        assert otu.accessions == set(isolate_1_accessions)

        isolate_1 = otu.isolates[0]

        assert isolate_1
        assert isolate_1 == otu.get_isolate(isolate_1.id)

        with precached_repo.lock():
            isolate_2 = add_and_name_isolate(
                precached_repo,
                otu,
                isolate_2_accessions,
                IsolateName(type=IsolateNameType.ISOLATE, value="dummy"),
                ncbi_client,
            )

        assert isolate_2

        otu_after = precached_repo.get_otu(otu.id)

        assert otu_after
        assert otu_after.isolate_ids == {isolate_1.id, isolate_2.id}

        assert isolate_2 == otu_after.get_isolate(isolate_2.id)
        assert isolate_2.accessions == set(isolate_2_accessions)
        assert isolate_2.name == IsolateName(
            type=IsolateNameType.ISOLATE,
            value="dummy",
        )

    def test_blocked(self, precached_repo: Repo):
        """Test that an isolate cannot be added to an OTU if both its name and its
        accessions are already contained.
        """
        accessions = ["MF062136", "MF062137", "MF062138"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu = otu_service.create(accessions)

            assert otu
            assert (
                add_genbank_isolate(precached_repo, otu, accessions, ncbi_client)
                is None
            )

    def test_name_conflict(self, precached_repo: Repo):
        """Test that an isolate cannot be added to an OTU if its name is already contained."""
        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(precached_repo, ncbi_client)

        with precached_repo.lock():
            otu_before = otu_service.create(["MF062136", "MF062137", "MF062138"])

        assert otu_before

        last_event_id_before_test = precached_repo.last_id

        with precached_repo.lock():
            isolate = add_and_name_isolate(
                precached_repo,
                otu_before,
                ["NC_055390", "NC_055391", "NC_055392"],
                IsolateName(type=IsolateNameType.ISOLATE, value="4342-5"),
                ncbi_client,
            )

        assert isolate

        otu_after = precached_repo.get_otu(otu_before.id)

        assert otu_after
        assert isolate.id not in otu_after.isolate_ids
        assert precached_repo.last_id == last_event_id_before_test

    def test_plan_mismatch(self, scratch_repo: Repo, mock_ncbi_client):
        """Test that an isolate cannot be added to if it does not match the plan."""
        taxid = mock_ncbi_client.otus.cabbage_leaf_curl_jamaica_virus.taxid
        otu_before = scratch_repo.get_otu_by_taxid(taxid)

        assert otu_before

        ncbi_client = NCBIClient(ignore_cache=True)

        with scratch_repo.lock():
            assert (
                add_genbank_isolate(
                    scratch_repo,
                    otu_before,
                    ["AB017503"],
                    ncbi_client,
                )
                is None
            )

        otu_after = scratch_repo.get_otu_by_taxid(taxid)

        assert otu_after
        assert otu_after == otu_before
        assert "AB017503" not in otu_after.accessions

    def test_auto_promote_refseq_monopartite(self, empty_repo: Repo):
        """Test that RefSeq accessions automatically promote GenBank sequences (monopartite)."""
        taxid = 1169032
        original_accessions = ["AB017503"]
        refseq_accessions = ["NC_003355"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(empty_repo, ncbi_client)

        with empty_repo.lock():
            otu = otu_service.create(accessions=original_accessions)

        assert otu
        assert otu.accessions == set(original_accessions)

        original_isolate = otu.isolates[0]
        original_isolate_name = original_isolate.name

        with empty_repo.lock():
            isolate = add_genbank_isolate(
                empty_repo, otu, refseq_accessions, ncbi_client
            )

        assert isolate is not None
        assert isolate.name == original_isolate_name
        assert isolate.accessions == set(refseq_accessions)

        otu_after = empty_repo.get_otu(otu.id)

        assert otu_after
        assert otu_after.accessions == set(refseq_accessions)
        assert set(original_accessions).issubset(otu_after.excluded_accessions)

    def test_auto_promote_refseq_multipartite(self, empty_repo: Repo, mock_ncbi_client):
        """Test that RefSeq accessions automatically promote GenBank sequences (multipartite)."""
        taxid = mock_ncbi_client.otus.cabbage_leaf_curl_jamaica_virus.taxid
        original_accessions = ["DQ178608", "DQ178609"]
        refseq_accessions = ["NC_038792", "NC_038793"]

        ncbi_client = NCBIClient(ignore_cache=True)
        otu_service = OTUService(empty_repo, ncbi_client)

        with empty_repo.lock():
            otu = otu_service.create(accessions=original_accessions)

        assert otu
        assert otu.accessions == set(original_accessions)

        original_isolate = otu.isolates[0]
        original_isolate_name = original_isolate.name

        with empty_repo.lock():
            isolate = add_genbank_isolate(
                empty_repo, otu, refseq_accessions, ncbi_client
            )

        assert isolate is not None
        assert isolate.name == original_isolate_name
        assert isolate.accessions == set(refseq_accessions)

        otu_after = empty_repo.get_otu(otu.id)

        assert otu_after
        assert otu_after.accessions == set(refseq_accessions)
        assert set(original_accessions).issubset(otu_after.excluded_accessions)
