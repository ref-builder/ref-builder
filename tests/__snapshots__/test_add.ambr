# serializer version: 1
# name: TestCreateOTU.test_empty_success
  dict({
    'acronym': '',
    'excluded_accessions': set({
    }),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'double',
        'topology': 'circular',
        'type': 'DNA',
      }),
      'multipartite': True,
      'segments': list([
        dict({
          'length': 2575,
          'name': 'DNA A',
          'required': True,
        }),
        dict({
          'length': 2494,
          'name': 'DNA B',
          'required': True,
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestCreateOTU.test_otu_create_refseq_autoexclude
  set({
    'EF546808',
    'EF546809',
    'EF546810',
    'EF546811',
    'EF546812',
    'EF546813',
  })
# ---
# name: TestCreateOTUCommands.test_autofill_ok
  dict({
    'molecule': dict({
      'strandedness': 'double',
      'topology': 'circular',
      'type': 'DNA',
    }),
    'multipartite': True,
    'segments': list([
      dict({
        'length': 2575,
        'name': 'DNA-A',
        'required': True,
      }),
      dict({
        'length': 2494,
        'name': 'DNA-B',
        'required': True,
      }),
    ]),
  })
# ---
# name: TestCreateOTUCommands.test_autofill_ok.1
  dict({
    'length': 2575,
    'name': 'DNA A',
    'required': True,
  })
# ---
# name: TestCreateOTUCommands.test_autofill_ok.2
  dict({
    'length': 2494,
    'name': 'DNA B',
    'required': True,
  })
# ---
# name: TestCreateOTUCommands.test_ok[1278205-accessions0]
  dict({
    'acronym': '',
    'excluded_accessions': set({
      'JX263426',
    }),
    'legacy_id': None,
    'name': 'Dahlia latent viroid',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'single',
        'topology': 'linear',
        'type': 'RNA',
      }),
      'multipartite': False,
      'segments': list([
        dict({
          'length': 342,
          'name': 'genomic RNA',
          'required': True,
        }),
      ]),
    }),
    'taxid': 1278205,
  })
# ---
# name: TestCreateOTUCommands.test_ok[345184-accessions1]
  dict({
    'acronym': '',
    'excluded_accessions': set({
    }),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'double',
        'topology': 'circular',
        'type': 'DNA',
      }),
      'multipartite': True,
      'segments': list([
        dict({
          'length': 2575,
          'name': 'DNA A',
          'required': True,
        }),
        dict({
          'length': 2494,
          'name': 'DNA B',
          'required': True,
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestUpdateOTU.test_without_exclusions_ok
  list([
    dict({
      'acronym': '',
      'excluded_accessions': list([
      ]),
      'legacy_id': None,
      'name': 'Apple rubbery wood virus 1',
      'schema': dict({
        'molecule': dict({
          'strandedness': 'single',
          'topology': 'linear',
          'type': 'RNA',
        }),
        'multipartite': True,
        'segments': list([
          dict({
            'length': 7219,
            'name': 'RNA L',
            'required': True,
          }),
          dict({
            'length': 1613,
            'name': 'RNA M',
            'required': True,
          }),
          dict({
            'length': 1291,
            'name': 'RNA S',
            'required': True,
          }),
        ]),
      }),
      'taxid': 2164102,
    }),
    IsolateName(type=<IsolateNameType.ISOLATE: 'isolate'>, value='4342-5'): set({
      'MF062136',
      'MF062137',
      'MF062138',
    }),
    IsolateName(type=<IsolateNameType.ISOLATE: 'isolate'>, value='982-11'): set({
      'NC_055390',
      'NC_055391',
      'NC_055392',
    }),
    IsolateName(type=<IsolateNameType.ISOLATE: 'isolate'>, value='BR-Mishima'): set({
      'MK936225',
      'MK936226',
      'MK936227',
    }),
    IsolateName(type=<IsolateNameType.ISOLATE: 'isolate'>, value='Pe2'): set({
      'OQ420743',
      'OQ420744',
      'OQ420745',
    }),
  })
# ---
